src/ledger-bridge-keyring.ts

```
@@ -0,0 +1,883 @@
+      const addressSignedWith = sigUtil.recoverTypedSignature_v4({
+        data,
+        sig: signature,
+      });
+      if (
+        ethUtil.toChecksumAddress(addressSignedWith) !==
+        ethUtil.toChecksumAddress(withAccount)
+      ) {
+        throw new Error('Ledger: The signature doesnt match the right address');
+      }
+      return signature;
+    }
+    throw (
+      payload.error || new Error('Ledger: Unknown error while signing message')
+    );
+  }
+
+  exportAccount() {
+    throw new Error('Not supported on this device');
+  }
+
+  forgetDevice() {
+    this.accounts = [];
+    this.page = 0;
+    this.unlockedAccount = 0;
+    this.paths = {};
+    this.accountDetails = {};
+    this.hdk = new HDKey();
+  }
+
+  /* PRIVATE METHODS */
+
+  #setupIframe() {
+    this.iframe = document.createElement('iframe');
+    this.iframe.src = this.bridgeUrl;
+    this.iframe.allow = `hid 'src'`;
+    this.iframe.onload = async () => {
+      // If the ledger live preference was set before the iframe is loaded,
+      // set it after the iframe has loaded
+      this.iframeLoaded = true;
+      if (this.delayedPromise) {
+        try {
+          const result = await this.updateTransportMethod(
+            this.delayedPromise.transportType,
+          );
+          this.delayedPromise.resolve(result);
+        } catch (e) {
+          this.delayedPromise.reject(e);
+        } finally {
+          delete this.delayedPromise;
+        }
+      }
+    };
+    document.head.appendChild(this.iframe);
+  }
+
+  #getOrigin() {
+    const tmp = this.bridgeUrl.split('/');
+    tmp.splice(-1, 1);
+    return tmp.join('/');
+  }
+
+  #sendMessage(
+    msg: Partial<IFrameMessage>,
+    cb: (response: IFrameMessageResponse) => void,
+  ) {
+    msg.target = 'LEDGER-IFRAME';
+
+    this.currentMessageId += 1;
+    msg.messageId = this.currentMessageId;
+
+    this.messageCallbacks[this.currentMessageId] = cb;
+    this.iframe?.contentWindow?.postMessage(msg, '*');
```

Instead of implicitly doing nothing if `this.iframe` has not been set yet, should we throw an error?

---

I added a check that throws an error if the iframe is undefined or it has not been loaded yet (so before `this.iframe.onload` has been called).